File: Solved/p10341.cpp
/*
*
* Problema UVA 10341 - Solve It
*
*/

#include <iostream>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

using namespace std;

int main(){
	double p, q, r, s, t, u;
	double val0, val1;

	double xi, xu, xr;
	double valXi, valXr;

	while( scanf("%lf %lf %lf %lf %lf %lf", &p, &q, &r, &s, &t, &u) == 6 ){
		val0 = p*exp(0) + q*sin(0) + r*cos(0) + s*tan(0) + t*0*0 + u;
		val1 = p*exp(-1) + q*sin(1) + r*cos(1) + s*tan(1) + t*1*1 + u;
		if( ( val0 * val1 ) > 0 ){
			printf("No solution\n");
		}else{
			/* Teorema de Bolzano o Teorema de Biseccion */
			xi = 0;
			xu = 1;
			while( xi + 0.0000000001 < xu ){
				xr = (xi+xu)/2;
				valXi = p*exp(-xi) + q*sin(xi) + r*cos(xi) + s*tan(xi) + t*xi*xi + u;
				valXr = p*exp(-xr) + q*sin(xr) + r*cos(xr) + s*tan(xr) + t*xr*xr + u;
				if( ( valXi * valXr ) <= 0 ){
					xu = xr;
				}else if( ( valXi * valXr ) > 0 ){
					xi = xr;
				}
				
			}
			printf("%.4lf\n",(xi+xu)/2);

		}
	}

  return 0;
}

File: Solved/p10341.py
'''
*
*
* Problema UVa 10341 - Solve It
*
*
'''

from math import *
from sys import stdin

epsilon = 0.0000000001

def f(p, q, r, s, t, u, x):
  return p*exp(-1*x)+q*sin(x) + r*cos(x) + s*tan(x) + t*x*x + u

def main():
  lines = stdin.readlines()
  out = ""
  for case in lines:
    case = list(map(int, case.strip().split() ))
    lo, hi = 0, 1
    xi = f(case[0], case[1], case[2], case[3], case[4], case[5], lo)
    xu = f(case[0], case[1], case[2], case[3], case[4], case[5], hi)

    if( xi*xu > 0 ):
      out += "No solution\n"
    else:
      while lo+epsilon < hi:
        mid = (lo+hi)/2
        valXi = f(case[0], case[1], case[2], case[3], case[4], case[5], lo)
        valXr = f(case[0], case[1], case[2], case[3], case[4], case[5], mid)
        if valXi*valXr <= 0:
          hi = mid
        else:
          lo = mid
      out += str( "{0:.4f}\n".format((lo+mid)/2) )
  print(out.strip())

main()